x/storage/keeper/payment.go

```diff
@@ -100,218 +148,353 @@ func (k Keeper) UnlockStoreFee(ctx sdk.Context, bucketInfo *storagetypes.BucketI
 	return nil
 }
 
-func (k Keeper) UnlockAndChargeStoreFee(ctx sdk.Context, bucketInfo *storagetypes.BucketInfo, objectInfo *storagetypes.ObjectInfo) error {
+func (k Keeper) UnlockAndChargeObjectStoreFee(ctx sdk.Context, bucketInfo *storagetypes.BucketInfo,
+	internalBucketInfo *storagetypes.InternalBucketInfo, objectInfo *storagetypes.ObjectInfo) error {
 	// unlock store fee
-	err := k.UnlockStoreFee(ctx, bucketInfo, objectInfo)
+	err := k.UnlockObjectStoreFee(ctx, bucketInfo, objectInfo)
 	if err != nil {
 		return fmt.Errorf("unlock store fee failed: %w", err)
 	}
 
-	return k.ChargeStoreFee(ctx, bucketInfo, objectInfo)
+	return k.ChargeObjectStoreFee(ctx, bucketInfo, internalBucketInfo, objectInfo)
+}
+
+func (k Keeper) IsPriceChanged(ctx sdk.Context, primarySpId uint32, priceTime int64) (bool, error) {
+	prePrice, err := k.paymentKeeper.GetStoragePrice(ctx, types.StoragePriceParams{
+		PrimarySp: primarySpId,
+		PriceTime: priceTime,
+	})
+	if err != nil {
+		return false, fmt.Errorf("get storage price failed: %w", err)
+	}
+	currentPrice, err := k.paymentKeeper.GetStoragePrice(ctx, types.StoragePriceParams{
+		PrimarySp: primarySpId,
+		PriceTime: ctx.BlockTime().Unix(),
+	})
+	if err != nil {
+		return false, fmt.Errorf("get storage price failed: %w", err)
+	}
+
+	return !(prePrice.ReadPrice.Equal(currentPrice.ReadPrice) &&
+		prePrice.PrimaryStorePrice.Equal(currentPrice.PrimaryStorePrice) &&
+		prePrice.SecondaryStorePrice.Equal(currentPrice.SecondaryStorePrice)), nil
 }
 
-func (k Keeper) ChargeStoreFee(ctx sdk.Context, bucketInfo *storagetypes.BucketInfo, objectInfo *storagetypes.ObjectInfo) error {
-	chargeSize, err := k.GetChargeSize(ctx, objectInfo.PayloadSize, objectInfo.CreateAt)
+func (k Keeper) ChargeObjectStoreFee(ctx sdk.Context, bucketInfo *storagetypes.BucketInfo,
+	internalBucketInfo *storagetypes.InternalBucketInfo, objectInfo *storagetypes.ObjectInfo) error {
+	chargeSize, err := k.GetObjectChargeSize(ctx, objectInfo.PayloadSize, objectInfo.CreateAt)
 	if err != nil {
 		return fmt.Errorf("get charge size error: %w", err)
 	}
-	return k.ChargeViaBucketChange(ctx, bucketInfo, func(bi *storagetypes.BucketInfo) error {
-		bi.BillingInfo.TotalChargeSize += chargeSize
-		secondarySpObjectsSize := bi.BillingInfo.SecondarySpObjectsSize
-		for _, sp := range objectInfo.SecondarySpAddresses {
-			secondarySpObjectsSize = append(secondarySpObjectsSize, storagetypes.SecondarySpObjectsSize{
-				SpAddress:       sp,
-				TotalChargeSize: chargeSize,
-			})
+
+	priceChanged, err := k.IsPriceChanged(ctx, bucketInfo.PrimarySpId, internalBucketInfo.PriceTime)
+	if err != nil {
+		return fmt.Errorf("check whether price changed error: %w", err)
+	}
+
+	if !priceChanged {
+		err := k.ChargeViaObjectChange(ctx, bucketInfo, internalBucketInfo, objectInfo, chargeSize, false)
```

Comment:
seems the payment flow of the object owner does not change.

Reply:
the flows are from bucket payment account to gvg & gvg family.
the payment account comes from the ower.